### PR TITLE
wip-mSHEC : multiple Shingled Erasure Code

### PR DESCRIPTION
**Pull request by t-miyamae**
blueprint of Shingled Erasure Code (SHEC)
https://wiki.ceph.com/Planning/Blueprints/Hammer/Shingled_Erasure_Code_(SHEC)

### CODE REVIEW

**Comment by t-miyamae** [comment]
I had modified Makefile.am in shec directory (r36).
I hope that would solve the bot make-check errors.

**Comment by dachary** [comment]
Make check passes, very good sign :-) Could you please squash the commits together and add a descriptive commit message ? The idea being that when someone looks at the history of the logs it is crystal clear what this commit does. If you'd like to split the commit into multiple commits that have different focus (like tests, updating the makefiles etc.) it's also fine. 

At the end of each commit message there must also be a line like
<pre>
Signed-off-by: Miyamae, Takeshi &lt;miyamae.takeshi@jp.fujitsu.com>
</pre>

**Comment by dachary** [comment]
Running the tests with valgrind as
<pre>
libtool --mode=execute valgrind --tool=memcheck ./unittest_erasure_code_shec
</pre>
shows the following error
<pre>
[ RUN      ] ErasureCodeShec.init_1
==30317== Conditional jump or move depends on uninitialised value(s)
==30317==    at 0x8EDF92: shec_calc_recovery_efficiency1(int, int, int, int, int) (shec.cc:64)
==30317==    by 0x8EE0FE: shec_reedsolomon_coding_matrix(int, int, int, int, int) (shec.cc:111)
==30317==    by 0x8E62A6: ErasureCodeShecReedSolomonVandermonde::prepare() (ErasureCodeShec.cc:389)
==30317==    by 0x8E24C3: ErasureCodeShec::init(std::map<std::string, std::string, std::less<std::string>, std::allocator<std::pair<std::string const, std::string> > > const&) (ErasureCodeShec.cc:71)
==30317==    by 0x896282: ErasureCodeShec_init_1_Test::TestBody() (TestErasureCodeShec.cc:46)
==30317==    by 0x91C64B: testing::Test::Run() (gtest.cc:2095)
==30317==    by 0x91CBB5: testing::internal::TestInfoImpl::Run() (gtest.cc:2314)
==30317==    by 0x91D0BE: testing::TestCase::Run() (gtest.cc:2420)
==30317==    by 0x9215BF: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:4024)
==30317==    by 0x920575: testing::UnitTest::Run() (gtest.cc:3687)
==30317==    by 0x8C91D2: main (TestErasureCodeShec.cc:3178)
==30317== 
</pre>
but the rest of the run is valgrind clean, congratulations :-)

**Comment by dachary** [comment]
@t-miyamae unfortunately the github web interface does not allow you to change the content of the commit message. You can only do that by rebasing ( see https://github.com/t-miyamae/mSHEC/commit/d0d7dde4a785923b9ab9593d80dca5cfbc895c92 : the commit message does not have your Signed-off-by).

You can squash the commits together with **git rebase -i HEAD~9** (i.e. the last 9 commits). In the editor that will open, you can change **pick** to squash for the commits you want to merge with the previous one. Once you have just one or two commits, it will be a lot easier to write a nice commit message.

File: src/erasure-code/shec/ErasureCodePluginShec.cc [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	   << dendl;
+      return -ENOENT;
+    }
+    int err = interface->init(parameters);
+    if (err) {
+      return err;
+    }
+    *erasure_code = ErasureCodeInterfaceRef(interface);
+
+    dout(0) << "ErasureCodePluginShec: factory() completed" << dendl;
```
Comment (dachary) [inline src/erasure-code/shec/ErasureCodePluginShec.cc]: dout(10) instead so that it does not always show

**Comment by dachary** [comment]
For the record, the non regression tests are worked on at https://github.com/t-miyamae/ceph-erasure-code-corpus/pull/1

**Comment by dachary** [comment]
@t-miyamae I see a merge in the most recent commits and I'm assuming you're working on rebasing. I'll wait until you let me know it's ready for review.

**Comment by dachary** [comment]
@t-miyamae it would be great if you could rebase the series against master. In the past 500 commits that have been merged since you started the series there is a file rename that makes it painful to switch from master to your branch (all needs to be recompiled because automake does not gracefully handle file renaming). Thanks in advance :-)

**Comment by t-miyamae** [comment]
I have updated all comments to include Signed-off-by.
Would you wait a minutes until I will commit some other updates, please?

**Comment by dachary** [comment]
@t-miyamae sure, I can wait, thanks !

**Comment by t-miyamae** [comment]
I supposed we fixed all problems except 2nd valgrind error(in ~ErasureCodeShecTableCache()) you sent. We are utilizing isa codes and the error occurs in case of isa as well. It seems a little hard for us to solve. Do you have any idea?

**Comment by dachary** [comment]
@t-miyamae I'll take a look at the TableCache valgrind error. If it's not yours no reason for it to be a blocker. 

Could you please rebase / squash the commits so that there are only a few (like 2 or 3) ? That makes it easier for review.

**Comment by t-miyamae** [comment]
That makes sense. I have just understood the meaning of squash. I will try to squash all the commits.
However, I'm not sure that I could reset merge commits in a right way. Anyway, I will try.

**Comment by t-miyamae** [comment]
If it's alright with you, we would like to have an option to submit another pull request instead of this (#3484).
Do we have to suffer any disadvantages from it ?

**Comment by dachary** [comment]
No disadvantage at all, feel free to submit another pull request. However, if the troube you are facing is that the git push is rejected after you rebased, this is expected: you should git push --force after a rebase. You can always find the previous set of commits from the reflog on your machine. Or when looking at the hash commit in the loic-bot messages : they point to the commit that was used when you last pushed (it is not otherwise archived on github which is annoying).

**Pull request closed without merging**